torchtune/training/checkpointing/_checkpointer.py

```diff
@@ -912,3 +923,182 @@ def save_checkpoint(
+            self._output_dir, f"{self._checkpoint_dir_prefix}_{epoch}"
+        )
+
+        if self._checkpoint_future is not None:
+            # Previous checkpoint needs to finish before saving the next one.
+            wait_start = time.perf_counter()
+
+            if self._is_rank_zero:
+                logger.info(
+                    "Previous checkpoint has not finished. Checkpointing frequency is too high. Waiting..."
+                )
+
+            self._checkpoint_future.result()
+
+            if self._is_rank_zero:
+                logger.info(
+                    f"Waited {time.perf_counter() - wait_start:.2f} seconds for previous checkpoint to finish"
+                )
+
+        cp_start = time.perf_counter()
+
+        if save_async:
+
+            def callback(
+                f: Future,
+            ) -> None:
+                if f.exception() is None:
+                    if self._is_rank_zero:
+                        logger.info(
+                            f"Checkpoint is saved asynchronously to {checkpoint_path}"
+                        )
+
+                    for index in range(epoch):
+                        last_checkpoint = os.path.join(
+                            self._checkpoint_dir,
+                            f"{self._checkpoint_dir_prefix}_{index}",
+                        )
+                        if os.path.isfile(
+                            os.path.join(last_checkpoint, self._metadata_file)
+                        ):
+                            if self._is_rank_zero:
+                                logger.info(
+                                    f"Cleaning up the last checkpoint {last_checkpoint} to save storage"
+                                )
+                            shutil.rmtree(last_checkpoint)
+                else:
+                    logger.error(
+                        f"Checkpoint failed to save asynchronously to {checkpoint_path} with the exception {f.exception()}"
+                    )
+
+            self._checkpoint_future = async_save(
+                state_dict,
+                storage_writer=FileSystemWriter(
+                    checkpoint_path,
+                    thread_count=16,
+                    single_file_per_rank=False,
+                    sync_files=False,
+                ),
+            )
+
+            if self._is_rank_zero:
+                logger.info(
+                    f"Trainer was blocked for {time.perf_counter() - cp_start:.2f} seconds for checkpointing to finish..."
+                )
+
+            self._checkpoint_future.add_done_callback(callback)
+        else:
+            if self._is_rank_zero:
+                logger.info(
+                    f"Saving model checkpoint synchronously to {checkpoint_path}."
+                )
+
+            save(
+                state_dict,
+                storage_writer=FileSystemWriter(
+                    checkpoint_path,
+                    thread_count=8,
```

Comment:
Out of curiosity, how is this value determined?

Reply:
This is set as per the available storage I/O bandwidth. This value determines the number of parallel I/O threads while writing the data to storage. For our internal storage like Manifold or even S3, we know the capacity available and tune this value accordingly. This way we use max IO capacity with optimal performance without getting throttled. 

For TorchTune case, since we are writing to Disk, we may have to run some experiments and tune it as per the IOPS available. Ideally, this should also be exposed as a checkpointing_config to the user. They can tune it to get the best performance depending on the underlying storage: HDD, SSD or cloud storage like S3.

I used 16 since that is a default for Manifold integrations internally. For Disks, this number can be higher.  Also, it will only affect the upload speed to storage so its virtually has no effect on the async checkpointing blocking time.